eos/effects/adaptivearmorhardener.py

```diff
@@ -2,17 +2,207 @@
+                elif (damagePattern_tuple[0][4] == 1) or (countPasses >= 15):
+                    # Most likely the RAH is cycling between two different profiles and is in an infinite loop.
+                    # Reduce the amount of resists we shift each loop, to try and stabilize on a more average profile.
+                    if resistanceShiftAmount > .01:
+                        resistanceShiftAmount = resistanceShiftAmount - .01
+                        logger.debug("Reducing resistance shift amount to %f", resistanceShiftAmount)
+                    else:
+                        # Start a new counter
+                        countPassesDos = countPassesDos+1
+
+                    countPasses = countPasses+1
+                    #logger.debug("Looped %f times.", countPasses)
+                else:
+                    countPasses = countPasses+1
+                    #logger.debug("Looped %f times.",countPasses)
+
+                # Loop through the resists that did least damage and figure out how much there is to steal
+                vampDmg=[0]*2
+                for i in [0,1]:
+                    attr = "armor%sDamageResonance" % damagePattern_tuple[i][0].capitalize()
+                    if damagePattern_tuple[i][4] < 1-resistanceShiftAmount:
+                        # If there is more than 3% to steal, let's steal 3% and reduce the resit by that amount.
+                        vampDmg[i] = resistanceShiftAmount
+                    else:
+                        # If there is equal to or less than 3% left to steal, grab what we can and set the resist to 0%.
+                        vampDmg[i] = 1-damagePattern_tuple[i][4]
+
+                    # Adjust the module resists down
+                    if vampDmg[i] > 0:
+                        module.increaseItemAttr(attr, vampDmg[i])
+                        # Set our "ship" resists
+                        damagePattern_tuple[i][1] = fit.ship.getModifiedItemAttr(attr) * module.getModifiedItemAttr(attr)
+
+                #Add up the two amounts we stole, and divide it equally between the two
+                vampDmgTotal = vampDmg[0] + vampDmg[1]
+
+                # Loop through the resists that took the most damage, and add what we vamped from the weakest resists.
+                for i in [2,3]:
+                    attr = "armor%sDamageResonance" % damagePattern_tuple[i][0].capitalize()
+                    # And by add we mean subtract, because in CCPland up is down and down is up
+                    if vampDmgTotal > 0:
+                        module.increaseItemAttr(attr, 0-(vampDmgTotal/2))
+                        # Set our "ship" resists
+                        damagePattern_tuple[i][1] = fit.ship.getModifiedItemAttr(attr)*module.getModifiedItemAttr(attr)
+
+            adaptiveResists_tuple = [0.0] * 12
+            for damagePatternType in damagePattern_tuple:
+                attr = "armor%sDamageResonance" % damagePatternType[0].capitalize()
```

Comment:
>Why the call to capitalize() when it's already capitalized?

To normalize data.

CCP is really bad about this, and it doesn't help that you'll find examples of this scattered around.

For example, you'll find `EM`, `Em`, and `em`, but for this particular case we _have_ to have `Em` or it won't read properly.

It's just a good coding practice.